Fix full dependencies that should be order only in ninja backend

dcbaker opened the pull request:
The current terminology is odd, it calls them "header deps" and "order deps", but headers are only order deps, and order deps are full dependencies. This renames and fixes the kind of dependencies being used.

bruchar1 commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -3130,6 +3125,17 @@ def add_dependency_scanner_entries_to_element(self, target: build.BuildTarget, c
         return os.path.join(self.get_target_private_dir(target), 'depscan.dd')
 
+    def _add_deps(self, target: build.BuildTarget, deps: T.Sequence[mesonlib.FileOrString],
+                  to: T.Callable[[str], None]) -> None:
```
`to` seems an odd name for a callable... What about something like `add_dep_to` ?

bruchar1 reviewed (commented):
I think the third commit message could be a little more explicit. Like "refactored by adding a `add_order_deps` function", or something like that.

dcbaker commented:
@bruchar1 done and done. Thanks for the review :)

bruchar1 reviewed (approved):
LGTM! Should it be added to the next point release?

dcbaker commented:
> Should it be added to the next point release?

I don't think so. This change potentially makes re-compiles faster when using generated headers, but it doesn't affect the correctness of the build

dcbaker committed:
backends/ninja: rename the add_header_deps method

header dependency are usually order-only, but this actually provides a
full dependency.

dcbaker committed:
backend/ninja: add docstring for generate_single_compile

dcbaker committed:
backends/ninja: fix dependencies added as full but should be order-only

Due to the confusing naming of the parameters, a number of arguments
were incorrectly added as full dependencies when they should be order
only.

bonzini commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -3242,15 +3243,23 @@ def order_deps_to_strings(self, target: build.BuildTarget, order_deps: T.List[Fi
     def generate_single_compile(self, target: build.BuildTarget, src: FileOrString,
                                 is_generated: bool = False,
-                                header_deps: T.Optional[T.List[FileOrString]] = None,
+                                full_deps: list[FileOrString] | list[File] | None = None,
```
You can use `T.Optional[FileList]`, or `FileList | None`:

- here
- in the `order_deps` line below
- in `order_deps_to_string`'s last argument

bonzini commented on mesonbuild/backend/ninjabackend.py:
Shouldn't this be `add_dep` now?

bonzini commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1146,7 +1146,7 @@ def generate_target(self, target: T.Union[build.Target]) -> None:
                 transpiled_source_files.append(raw_src)
         for src in transpiled_source_files:
-            o, s = self.generate_single_compile(target, src, True, [], header_deps)
+            o, s = self.generate_single_compile(target, src, True, full_deps=header_deps)
```
Why do these have to become full dependencies? Using order dependencies as before works because the order dependencies work in the first build and the depfile adds full dependencies for the rebuilds.

bonzini commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -3594,7 +3610,7 @@ def generate_pch(self, target: build.BuildTarget, header_deps: T.Optional[T.List
             if extradep is not None:
                 elem.add_dep(extradep)
-            self.add_header_deps(target, elem, header_deps)
+            self.add_full_deps(target, elem, header_deps)
```
This also seems incorrect, order dependency would be good.

bonzini reviewed (changes requested):
This was already fixed in 07f595fa98 (PR #14180). Leaving open as a naming cleanup, but applying this PR as is would change `add_full_deps` to add order dependencies
